File: app/src/main/java/pl/dawidraszka/compassproject/view/PositionDialogFragment.java
package pl.dawidraszka.compassproject.view;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import pl.dawidraszka.compassproject.R;
import pl.dawidraszka.compassproject.model.data.SimplePosition;

public class PositionDialogFragment extends DialogFragment {

    private PositionDialogListener listener;
    private Button okButton;
    private EditText latitudeEditText;
    private EditText longitudeEditText;

    private SimplePosition destination;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        listener = (PositionDialogListener) context;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        LayoutInflater inflater = requireActivity().getLayoutInflater();

        View view = inflater.inflate(R.layout.dialog_set_destination, null);

        okButton = view.findViewById(R.id.ok_button);

        view.findViewById(R.id.cancel_button).setOnClickListener(v -> getDialog().cancel());

        TextWatcher textChangedListener = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                textUpdated();
            }
        };

        latitudeEditText = view.findViewById(R.id.latitude_et);
        longitudeEditText = view.findViewById(R.id.longitude_et);

        latitudeEditText.addTextChangedListener(textChangedListener);
        longitudeEditText.addTextChangedListener(textChangedListener);

        okButton.setOnClickListener(v -> {
            listener.onDialogPositiveClick(destination);
            this.dismiss();
        });

        builder.setView(view)
                .setTitle(R.string.set_destination);
        return builder.create();
    }

    private void textUpdated() {
        double latitude;
        double longitude;

        try {
            latitude = Double.parseDouble(latitudeEditText.getText().toString());
            longitude = Double.parseDouble(longitudeEditText.getText().toString());

            if (latitudeCorrect(latitude) && longitudeCorrect(longitude)) {
                correctDestination(new SimplePosition(latitude, longitude));
            } else {
                incorrectDestination();
            }

        } catch (NumberFormatException e) {
            incorrectDestination();
        }
    }

    private boolean latitudeCorrect(double latitude) {
        return latitude >= -90 && latitude <= 90;
    }

    private boolean longitudeCorrect(double longitude) {
        return longitude >= -180 && longitude <= 180;
    }

    private void correctDestination(SimplePosition destination) {
        this.destination = destination;
        okButton.setEnabled(true);
    }

    private void incorrectDestination() {
        destination = null;
        okButton.setEnabled(false);
    }

    interface PositionDialogListener {
        void onDialogPositiveClick(SimplePosition destination);
    }
}

File: app/src/main/java/pl/dawidraszka/compassproject/view/PositionDialogFragment.kt
package pl.dawidraszka.compassproject.view

import android.app.Activity
import android.app.Dialog
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import kotlinx.android.synthetic.main.dialog_set_destination.*
import kotlinx.android.synthetic.main.dialog_set_destination.view.*
import pl.dawidraszka.compassproject.R
import pl.dawidraszka.compassproject.model.SimplePosition

class PositionDialogFragment : DialogFragment() {

    internal lateinit var listener: PositionDialogListener

    interface PositionDialogListener {
        fun onDialogPositiveClick(destination: SimplePosition)
    }

    override fun onAttach(activity: Activity) {
        super.onAttach(activity)
        listener = activity as PositionDialogListener
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return activity?.let {
            // Use the Builder class for convenient dialog construction
            val builder = AlertDialog.Builder(it)
            val inflater = requireActivity().layoutInflater

            val layout = inflater.inflate(R.layout.dialog_set_destination, null)

            builder.setView(layout)
                .setMessage(R.string.set_destination)
                .setPositiveButton(
                    R.string.ok
                ) { dialog, _ ->
                    listener.onDialogPositiveClick(SimplePosition(layout.latitude_et.text.toString().toDouble(), layout.longitude_et.text.toString().toDouble()))
                }
                .setNegativeButton(R.string.cancel) { dialog, _ ->
                    dialog.cancel()
                }
            // Create the AlertDialog object and return it
            builder.create()
        } ?: throw IllegalStateException("Activity cannot be null")
    }
}